### PR TITLE
Update `\yii\base\ErrorHandler`

### PR DESCRIPTION
**Pull request by WinterSilence**
1. Adds property `_workingDirectory` to restore current working directory in shutdown handler:
> `register_shutdown_function()`:
> Working directory of the script can change inside the shutdown function under some Web servers, e.g. Apache.
2. Adds event to call user's handler on application shutdown

| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️/❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  |

### CODE REVIEW

File: framework/base/ErrorHandler.php [hunk 1]
```diff
@@ -262,7 +275,12 @@ public function handleError($code, $message, $file, $line)
      */
     public function handleFatalError()
     {
-        unset($this->_memoryReserve);
+        $this->_memoryReserve = null;
```
Comment (samdark) [inline framework/base/ErrorHandler.php:278]: As far as I remember it works differently memory-wise. Needs to be re-checked.

File: framework/base/ErrorHandler.php [hunk 2]
```diff
@@ -262,7 +275,12 @@ public function handleError($code, $message, $file, $line)
...
+
+        // fix working directory for some Web servers
+        if (!in_array(PHP_SAPI, ['cli', 'phpdbg'], true)) {
```
Comment (samdark) [inline framework/base/ErrorHandler.php:281]: When/why does it matter?

What SAPIs are changing working dir during shutdown?

Comment (WinterSilence) [reply framework/base/ErrorHandler.php:281]: @samdark see PR description

File: framework/base/ErrorHandler.php (hunk 1 above)
Comment (WinterSilence) [reply framework/base/ErrorHandler.php:278]: @samdark i'm think is more correctly way

File: framework/base/ErrorHandler.php (hunk 2 above)
Comment (samdark) [reply framework/base/ErrorHandler.php]: https://www.php.net/manual/en/function.register-shutdown-function.php

File: framework/base/ErrorHandler.php (hunk 1 above)
Comment (samdark) [reply framework/base/ErrorHandler.php:278]: Not in this case. Setting null does not remove variable from the symbol table thus freeing a bit less memory. Let's revert to `unset()`. See https://stackoverflow.com/questions/584960/whats-better-at-freeing-memory-with-php-unset-or-var-null

**Comment by samdark** [comment]
- Overall cwd fix is 👍
- Unset change to null is 👎 Let's revert it.
- Need a line for CHANGELOG.

File: framework/base/ErrorHandler.php (hunk 1 above)
Comment (WinterSilence) [reply framework/base/ErrorHandler.php:278]: @samdark it's not variable, it's property

**Comment by WinterSilence** [comment]
@samdark я удивлен? что мне приходится объяснять основы: `unset()` удаляет свойство, при обращении к удаленному свойству будет генерироваться ошибка т.к. свойство удалено, а геттер отсутствует.

File: framework/base/ErrorHandler.php (hunk 1 above)
Comment (samdark) [reply framework/base/ErrorHandler.php:278]: Yes. That's the goal. Unset will remove it from symbol table freeing up a bit more memory.

Comment (samdark) [reply framework/base/ErrorHandler.php:278]: There will be no second access to this property.

Comment (WinterSilence) [reply framework/base/ErrorHandler.php:278]: @samdark economy 30-40 bytes? no reasons to use ugly solutions like this - yii have many unoptimized places

File: framework/base/ErrorHandler.php [hunk 3]
```diff
@@ -262,7 +275,10 @@ public function handleError($code, $message, $file, $line)
      */
     public function handleFatalError()
     {
-        unset($this->_memoryReserve);
+        $this->_memoryReserve = null;
+
+        // fix working directory for some Web servers e.g. Apache
+        chdir($this->_workingDirector);
```
Comment (samdark) [inline framework/base/ErrorHandler.php:281]:
```suggestion
        chdir($this->_workingDirector);
        unset($this->_workingDirector);
```

File: framework/base/ErrorHandler.php [hunk 4]
```diff
@@ -262,7 +275,11 @@ public function handleError($code, $message, $file, $line)
      */
     public function handleFatalError()
     {
-        unset($this->_memoryReserve);
+        // fix working directory for some Web servers e.g. Apache
+        chdir($this->_workingDirector);
```
Comment (samdark) [inline framework/base/ErrorHandler.php:279]:
```suggestion
        chdir($this->_workingDirectory);
```

File: framework/base/ErrorHandler.php [hunk 5]
```diff
@@ -262,7 +275,11 @@ public function handleError($code, $message, $file, $line)
...
+        // flush memory
+        $this->_workingDirector = null;
```
Comment (samdark) [inline framework/base/ErrorHandler.php:281]:
```suggestion
        unset($this->_workingDirectory);
```

File: framework/base/ErrorHandler.php [hunk 6]
```diff
@@ -262,7 +275,11 @@ public function handleError($code, $message, $file, $line)
...
+        $this->_memoryReserve = null;
```
Comment (samdark) [inline framework/base/ErrorHandler.php:282]:
```suggestion
        unset($this->_memoryReserve);
```

**Review by samdark — CHANGES_REQUESTED** [review]

**Comment by bizley** [comment]
Are we wrapping the last exit in register_shutdown_function as in the #19401 as well or is it an overkill?

**Comment by WinterSilence** [comment]
@bizley @samdark saves every byte, so he won't approve it (:

File: framework/base/ErrorHandler.php [hunk 7]
```diff
@@ -264,6 +281,13 @@ public function handleFatalError()
     {
         unset($this->_memoryReserve);
 
+        if (isset($this->_workingDirector)) {
+            // fix working directory for some Web servers e.g. Apache
+            chdir($this->_workingDirectory);
+            // flush memory
+            unset($this->_workingDirector);
```
Comment (cebe) [inline framework/base/ErrorHandler.php:288]:
```suggestion
            $this->_workingDirectory = null;
```

unset will actually remove the property from the class, which is different from setting it to `null`.

**Review by cebe — CHANGES_REQUESTED** [review]

File: framework/base/ErrorHandler.php [hunk 8]
```diff
@@ -83,7 +94,11 @@ public function register()
                 set_error_handler([$this, 'handleError']);
             }
             if ($this->memoryReserveSize > 0) {
-                $this->_memoryReserve = str_repeat('x', $this->memoryReserveSize);
+                $this->_memoryReserve = str_pad('x', $this->memoryReserveSize -1);
```
[the remark this answers is not in the document]
Comment (WinterSilence) [reply framework/base/ErrorHandler.php:97]: @samdark str_pad faster than str_repeat

[the remark this answers is not in the document]
Comment (WinterSilence) [reply framework/base/ErrorHandler.php]: @samdark fixed

File: framework/base/ErrorHandler.php (hunk 7 above)
Comment (bizley) [inline framework/base/ErrorHandler.php:288]:
```suggestion
            unset($this->_workingDirectory);
```

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
Thanks!

**Pull request merged**